internal/rules/resource_management.go

```diff
@@ -710,156 +669,154 @@ func (r ResourceManagementService) shouldDeleteServiceAccount(
 	return !namespaceInScope || !nameExpected
 }
 
-func (r ResourceManagementService) markServiceAccountForDeletion(
-	ctx context.Context,
-	sa *corev1.ServiceAccount,
-	targetNamespaces *set.Set[string],
-) error {
-	logger := log.FromContext(ctx).WithName("markServiceAccountForDeletion")
+func (r ResourceManagementService) GarbageCollectRoles(ctx context.Context, resources []WSAResource) error {
+	logger := log.FromContext(ctx).WithName("garbageCollectRoles")
 
-	namespaceInScope := targetNamespaces.Contains(sa.Namespace)
-	logger.Info("Marking ServiceAccount for deletion",
-		"name", sa.Name,
-		"namespace", sa.Namespace,
-		"reason", fmt.Sprintf("namespaceInScope=%v", namespaceInScope),
-		"annotations", sa.Annotations)
-
-	if deleteErr := r.client.Delete(ctx, sa); deleteErr != nil {
-		if !errors.IsNotFound(deleteErr) {
-			logger.Error(deleteErr, "failed to initiate deletion of ServiceAccount",
-				"name", sa.Name, "namespace", sa.Namespace)
-			return fmt.Errorf("failed to initiate deletion of ServiceAccount %s in namespace %s: %w",
-				sa.Name, sa.Namespace, deleteErr)
+	expectedRoles := set.New[string](len(resources))
```

Comment:
This could get clashes with just the name

```suggestion
	expectedRoles := set.New[NamespacedName](len(resources))
```